linker.ts

```diff
@@ -1,72 +1,133 @@
+function libraryHashPlaceholder (fullyQualifiedLibraryName) {
+  return `$${keccak256(fullyQualifiedLibraryName).slice(0, 34)}$`;
 }
 
-const linkBytecode = function (bytecode, libraries) {
+/**
+ * Used to support legacy library names and the current hash library names,
+ * replaces the names with the provided hex address.
+ *
+ * @param bytecode
+ * @param address
+ * @param label
+ */
+function replacePlaceholder (bytecode, label, address) {
+  // truncate to 36 characters
+  const truncatedName = label.slice(0, 36);
+  const libLabel = `__${truncatedName.padEnd(36, '_')}__`;
+
+  while (bytecode.indexOf(libLabel) >= 0) {
+    bytecode = bytecode.replace(libLabel, address);
+  }
+
+  return bytecode;
+}
+
+/**
+ * Finds and all library placeholders in the provided bytecode and replaces them with actual addresses.
+ * Supports both old- and new-style placeholders (even both in the same file).
+ * See [Library Linking](https://docs.soliditylang.org/en/latest/using-the-compiler.html#library-linking)
+ * for a full explanation of the linking process.
+ *
+ * Example of a legacy placeholder: `__lib.sol:L_____________________________`
+ * Example of a new-style placeholder: `__$cb901161e812ceb78cfe30ca65050c4337$__`
+ *
+ * @param bytecode Hex-encoded bytecode string. All 40-byte substrings starting and ending with
+ *     `__` will be interpreted as placeholders.
+ *
+ * @param libraries Mapping between fully qualified library names and the hex-encoded
+ *     addresses they should be replaced with. Addresses shorter than 40 characters are automatically padded with zeros.
+ *
+ * @returns bytecode Hex-encoded bytecode string with placeholders replaced with addresses.
+ *    Note that some placeholders may remain in the bytecode if `libraries` does not provide addresses for all of them.
+ */
+function linkBytecode (bytecode, libraries) {
   assert(typeof bytecode === 'string');
   assert(typeof libraries === 'object');
+
   // NOTE: for backwards compatibility support old compiler which didn't use file names
   const librariesComplete = {};
-  for (const libraryName in libraries) {
-    if (typeof libraries[libraryName] === 'object') {
-      // API compatible with the standard JSON i/o
-      for (const lib in libraries[libraryName]) {
-        librariesComplete[lib] = libraries[libraryName][lib];
-        librariesComplete[libraryName + ':' + lib] = libraries[libraryName][lib];
-      }
-    } else {
-      // backwards compatible API for early solc-js versions
-      const parsed = libraryName.match(/^([^:]+):(.+)$/);
-      if (parsed) {
-        librariesComplete[parsed[2]] = libraries[libraryName];
+
+  for (const [libraryName, library] of Object.entries(libraries)) {
+    if (isNil(library)) continue;
+
+    // API compatible with the standard JSON i/o
+    if (isObject(library)) {
+      for (const [key, value] of Object.entries(library)) {
+        librariesComplete[key] = value;
+        librariesComplete[`${libraryName}:${key}`] = value;
```

Comment:
` const sourceUnitName = libraryName;` I disagree on this. Having this `{"lib.sol": {"L": "0x..."}}` as a comment provides more context than an ambiguous variable name change.

Reply:
I have made minor changes here to support this where applicable.

Reply:
> I disagree on this. Having this `{"lib.sol": {"L": "0x..."}}` as a comment provides more context than an ambiguous variable name change.

It's not ambiguous at all - [*source unit name* has a very clear definition in this context](https://docs.soliditylang.org/en/latest/path-resolution.html#virtual-filesystem). It's `libraryName` that is ambiguous here, which is why I'm suggesting a rename. Having an additional comment here as an illustration won't hurt but it does not mean we can't have precise names at the same time :)

For me the only downside is that renaming a variable does not seem like a great style. I was actually interested to see what your opinion would be. But overall I think that it's a slight improvement over the old ambiguous name.

Reply:
But anyway, I'd say that names in the PR overall are good enough after your last changes, so if you disagree, I'd approve even without this change. I just wanted to say that think this would be a clearer name.

Reply:
I did directly go with this pathing but i did make some changes to make it more defined which path later on.